File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/geo/GeoDtos.kt
package app.pantopus.android.data.api.models.geo

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/** Normalized address from reverse geocode. Route: `backend/routes/geo.js`. */
@JsonClass(generateAdapter = true)
data class NormalizedAddress(
    val address: String? = null,
    val city: String? = null,
    val state: String? = null,
    val zipcode: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    @Json(name = "place_id") val placeId: String? = null,
    val verified: Boolean? = null,
    val source: String? = null,
) {
    val localityLabel: String
        get() =
            listOfNotNull(city, state)
                .map { it.trim() }
                .filter { it.isNotEmpty() }
                .joinToString(", ")
}

/** `GET /api/geo/reverse` envelope. */
@JsonClass(generateAdapter = true)
data class GeoReverseResponse(
    val normalized: NormalizedAddress,
)

/**
 * One address-typeahead suggestion from `GET /api/geo/autocomplete`.
 *
 * NOTE the wire shape (verified against the live backend): `center` is
 * a GeoJSON-style `[lng, lat]` ARRAY, not a `{lat, lng}` object (the
 * web `GeoSuggestion` TS type claims an object — the wire wins).
 */
@JsonClass(generateAdapter = true)
data class GeoSuggestion(
    @Json(name = "suggestion_id") val suggestionId: String,
    @Json(name = "place_id") val placeId: String? = null,
    @Json(name = "primary_text") val primaryText: String,
    @Json(name = "secondary_text") val secondaryText: String? = null,
    val label: String,
    val text: String? = null,
    /** GeoJSON order: `[longitude, latitude]`. */
    val center: List<Double> = emptyList(),
    val kind: String,
) {
    val longitude: Double? get() = center.getOrNull(0)
    val latitude: Double? get() = center.getOrNull(1)
}

/** `GET /api/geo/autocomplete` envelope. */
@JsonClass(generateAdapter = true)
data class GeoAutocompleteResponse(
    val suggestions: List<GeoSuggestion> = emptyList(),
)

/** `POST /api/geo/resolve` body. */
@JsonClass(generateAdapter = true)
data class GeoResolveRequest(
    @Json(name = "suggestion_id") val suggestionId: String,
)

/** `POST /api/geo/resolve` envelope. */
@JsonClass(generateAdapter = true)
data class GeoResolveResponse(
    val normalized: NormalizedAddress,
)

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Geo/GeoDTOs.swift
//
//  GeoDTOs.swift
//  Pantopus
//
//  DTOs for `/api/geo/*`. Route: `backend/routes/geo.js`.
//

import Foundation

/// Normalized address from reverse geocode / resolve.
public struct NormalizedAddress: Decodable, Sendable, Hashable {
    public let address: String?
    public let city: String?
    public let state: String?
    public let zipcode: String?
    public let latitude: Double?
    public let longitude: Double?
    public let placeId: String?
    public let verified: Bool?
    public let source: String?

    private enum CodingKeys: String, CodingKey {
        case address, city, state, zipcode, latitude, longitude, verified, source
        case placeId = "place_id"
    }

    /// City, state label for compose summaries.
    public var localityLabel: String {
        [city, state].compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .joined(separator: ", ")
    }
}

/// `GET /api/geo/reverse` envelope.
public struct GeoReverseResponse: Decodable, Sendable, Hashable {
    public let normalized: NormalizedAddress
}

/// One address-typeahead suggestion from `GET /api/geo/autocomplete`.
///
/// NOTE the wire shape (verified against the live backend): `center`
/// is a GeoJSON-style `[lng, lat]` ARRAY, not a `{lat, lng}` object
/// (the web `GeoSuggestion` TS type claims an object — the wire wins).
public struct GeoSuggestion: Decodable, Sendable, Hashable, Identifiable {
    public let suggestionId: String
    public let placeId: String?
    public let primaryText: String
    public let secondaryText: String?
    public let label: String
    public let text: String?
    /// GeoJSON order: `[longitude, latitude]`.
    public let center: [Double]
    public let kind: String

    public var id: String { suggestionId }

    public var longitude: Double? { center.count >= 2 ? center[0] : nil }
    public var latitude: Double? { center.count >= 2 ? center[1] : nil }

    private enum CodingKeys: String, CodingKey {
        case label, center, kind, text
        case suggestionId = "suggestion_id"
        case placeId = "place_id"
        case primaryText = "primary_text"
        case secondaryText = "secondary_text"
    }
}

/// `GET /api/geo/autocomplete` envelope.
public struct GeoAutocompleteResponse: Decodable, Sendable, Hashable {
    public let suggestions: [GeoSuggestion]
}

/// `POST /api/geo/resolve` body.
public struct GeoResolveRequest: Encodable, Sendable, Hashable {
    public let suggestionId: String

    public init(suggestionId: String) {
        self.suggestionId = suggestionId
    }

    private enum CodingKeys: String, CodingKey {
        case suggestionId = "suggestion_id"
    }
}

/// `POST /api/geo/resolve` envelope.
public struct GeoResolveResponse: Decodable, Sendable, Hashable {
    public let normalized: NormalizedAddress
}
